File: 3518-smallest-palindromic-rearrangement-ii/3518-smallest-palindromic-rearrangement-ii.cpp
class Solution {
    long long nCr(int n, int r, int k) {
        long long res = 1;
        r = min(r, n - r);
        for (int i = 1; i <= r; i++) {
            res = res * (n - i + 1) / i;
            if (res > k)
                return k + 1LL;
        }
        return res;
    }

    long long ways(int n, vector<int>& f, int k) {
        long long total = 1;
        for (int i = 0; i < 26; i++) {
            total *= nCr(n, f[i], k);
            if (total > k)
                return k + 1LL;
            n -= f[i];
        }
        return total;
    }

public:
    string smallestPalindrome(string s, int k) {
        int n = s.size(), len = n / 2;
        vector<int> f(26, 0);
        for (char c : s)
            f[c - 'a']++;

        string str(n, ' ');
        // middle char
        for (int i = 0; i < 26; i++) {
            if (f[i] & 1)
                str[n / 2] = char('a' + i);
            f[i] /= 2;
        }

        long long total = ways(len, f, k);
        if (total < k)
            return "";

        for (int idx = 0; idx < len; idx++) {
            for (int i = 0; i < 26; i++) {
                if (f[i] == 0)
                    continue;
                f[i]--;
                long long possible = ways(len - idx - 1, f, k);
                if (possible >= k) {
                    str[idx] = char('a' + i);
                    break;
                } else {
                    k -= possible;
                    f[i]++;
                }
            }
        }

        for (int i = 0; i < len; i++) {
            str[n - 1 - i] = str[i];
        }
        return str;
    }
};

File: 3518-smallest-palindromic-rearrangement-ii/3518-smallest-palindromic-rearrangement-ii.py
class Solution:
    def smallestPalindrome(self, s: str, k: int) -> str:
        n = len(s)
        half = n // 2
        freq = [0] * 26
        for ch in s:
            freq[ord(ch) - 97] += 1

        res = [' '] * n
        # middle character
        for i in range(26):
            if freq[i] & 1:
                res[n // 2] = chr(97 + i)
            freq[i] //= 2

        def nCr(n: int, r: int, limit: int) -> int:
            r = min(r, n - r)
            ans = 1
            for i in range(1, r + 1):
                ans = ans * (n - i + 1) // i
                if ans > limit:
                    return limit + 1
            return ans

        def ways(rem: int, f: list, limit: int) -> int:
            total = 1
            for i in range(26):
                total *= nCr(rem, f[i], limit)
                if total > limit:
                    return limit + 1
                rem -= f[i]
            return total

        total = ways(half, freq, k)
        if total < k:
            return ""

        for idx in range(half):
            for i in range(26):
                if freq[i] == 0:
                    continue
                freq[i] -= 1
                possible = ways(half - idx - 1, freq, k)
                if possible >= k:
                    res[idx] = chr(97 + i)
                    break
                else:
                    k -= possible
                    freq[i] += 1

        for i in range(half):
            res[n - 1 - i] = res[i]
        return ''.join(res)